### PR TITLE
Var interface value

### PR DESCRIPTION
**Pull request by antJack**
### Issues associated with this PR

Your PR should present related issues you want to solve.

### Solutions
You should show your solutions about the issues in your PR, including the overall solutions, details and the changes. At this time, Chinese is allowed to describe these.

### UT result
Unit Test is needed if the code is changed, your unit test should cover boundary cases, corner cases, and some exceptional cases. And you need to show the UT result.

### Benchmark
If your code involves the processing of every request, you should give the Benchmark Result.

### Code Style
+ Make sure `Goimports` has run
+ Show `Golint` result

### CODE REVIEW

File: pkg/variable/api.go [hunk 1]
```diff
@@ -27,19 +27,38 @@ import (
 )
 
 func GetVariableValue(ctx context.Context, name string) (string, error) {
+	v, err := GetVariableValueInterface(ctx, name)
+	if err != nil {
```
Comment (taoyuanyuan) [inline pkg/variable/api.go:31]: if {
return 
}
if {
 return
}

File: pkg/variable/var.go [hunk 2]
```diff
@@ -95,3 +126,41 @@ func BasicSetter(ctx context.Context, variableValue *IndexedValue, value string)
...
+	return nil
+}
+
+type basicSetter struct {
+	stringSetter    SetterFunc
+	interfaceSetter SetterFuncInterface
+}
+
+func (b *basicSetter) Set(ctx context.Context, variableValue *IndexedValue, value interface{}) error {
+	if b.stringSetter != nil {
```
Comment (taoyuanyuan) [inline pkg/variable/var.go:142]: 同上

File: pkg/variable/var.go [hunk 3]
```diff
@@ -95,3 +130,46 @@ func BasicSetter(ctx context.Context, variableValue *IndexedValue, value string)
 	variableValue.Valid = true
 	return nil
 }
+
+func BasicInterfaceSetter(ctx context.Context, variableValue *IndexedValue, value interface{}) error {
```
Comment (taoyuanyuan) [inline pkg/variable/var.go:134]: interface都不要了

File: pkg/variable/api.go [hunk 4]
```diff
@@ -78,7 +113,7 @@ func SetVariableValue(ctx context.Context, name, value string) error {
 }
 
 // TODO: provide direct access to this function, so the cost of variable name finding could be optimized
-func getFlushedVariableValue(ctx context.Context, index uint32) (string, error) {
+func getFlushedVariableValue(ctx context.Context, index uint32) (interface{}, error) {
```
Comment (taoyuanyuan) [inline pkg/variable/api.go:116]: getFlushedValue ？

File: pkg/variable/api.go [hunk 5]
```diff
@@ -99,7 +134,7 @@ func getFlushedVariableValue(ctx context.Context, index uint32) (string, error)
 	return "", errors.New(errNoVariablesInContext)
 }
 
-func getIndexedVariableValue(ctx context.Context, value *IndexedValue, index uint32) (string, error) {
+func getIndexedVariableValue(ctx context.Context, value *IndexedValue, index uint32) (interface{}, error) {
```
Comment (taoyuanyuan) [inline pkg/variable/api.go:137]: getIndexedValue

File: pkg/variable/api.go [hunk 6]
```diff
@@ -124,7 +159,7 @@ func getIndexedVariableValue(ctx context.Context, value *IndexedValue, index uin
 	return value.data, nil
 }
 
-func setFlushedVariableValue(ctx context.Context, index uint32, value string) error {
+func setFlushedVariableValue(ctx context.Context, index uint32, value interface{}) error {
```
Comment (taoyuanyuan) [inline pkg/variable/api.go:162]: setFlushedValue

File: pkg/variable/var.go [hunk 7]
```diff
@@ -65,33 +68,113 @@ func (iv *IndexedVariable) GetIndex() uint32 {
...
+			setter: &setterImpl{stringSetter: setter},
 			name:   name,
 			data:   data,
 			flags:  flags,
 		},
 	}
 }
 
-// BasicSetter used for variable value setting only, and would not affect any real data structure, like headers.
+// Deprecated: use DefaultStringSetter or DefaultSetter instead
```
Comment (taoyuanyuan) [inline pkg/variable/var.go:127]: deprecated的要不要都单独放个文件？

File: pkg/variable/var.go [hunk 8]
```diff
@@ -65,33 +68,113 @@ func (iv *IndexedVariable) GetIndex() uint32 {
...
 func BasicSetter(ctx context.Context, variableValue *IndexedValue, value string) error {
+	return DefaultStringSetter(ctx, variableValue, value)
+}
+
+// DefaultStringSetter used for string-typed variable value setting only, and would not affect any real data structure, like headers.
+func DefaultStringSetter(ctx context.Context, variableValue *IndexedValue, value string) error {
```
Comment (taoyuanyuan) [inline pkg/variable/var.go:133]: DefaultStringSetter和DefaultSetter一样的，是不是只需要DefaultSetter就行了？

Comment (antJack) [reply pkg/variable/var.go:133]: 最后一个参数类型不一样，一个是 string 一个是 interface，DefaultStringSetter 还是为了兼容旧代码，暂时去不掉

Comment (taoyuanyuan) [reply pkg/variable/var.go:133]: 那直接调用 DefaultSetter 就行了

File: pkg/variable/factory.go [hunk 9]
```diff
@@ -43,21 +43,12 @@ var (
 	errSupportIndexedOnly   = "this operation only support indexed variable"
 	errGetterNotFound       = "getter function undefined, variable name: "
 	errSetterNotFound       = "setter function undefined, variable name: "
+	errVariableNotString    = "variable type is not string, name: "
+	errValueNotString       = "set string variable with non-string type"
 )
 
-// ResetVariableForTest is a test function for reset the variables.
-// DONOT call it in any non-test functions
-func ResetVariableForTest() {
```
Comment (taoyuanyuan) [inline pkg/variable/factory.go:50]: @nejisama  这个还在用吗

File: pkg/variable/var.go [hunk 10]
```diff
@@ -65,33 +112,39 @@ func (iv *IndexedVariable) GetIndex() uint32 {
...
-	return &BasicVariable{
-		getter: getter,
-		setter: setter,
-		name:   name,
-		data:   data,
-		flags:  flags,
-	}
+type setterImpl struct {
+	stringSetter    StringSetterFunc
+	interfaceSetter SetterFunc
```
Comment (taoyuanyuan) [inline pkg/variable/var.go:117]: interfaceSetter -> setter  or  Setter ?

**Review by taoyuanyuan — APPROVED** [review]
LGTM

File: pkg/variable/factory.go (hunk 9 above)
Comment (nejisama) [reply pkg/variable/factory.go:50]: 用的

File: pkg/variable/var.go [hunk 11]
```diff
@@ -17,12 +17,59 @@
...
+// DefaultStringSetter used for string-typed variable value setting only, and would not affect any real data structure, like headers.
+func DefaultStringSetter(ctx context.Context, variableValue *IndexedValue, value string) error {
+	return DefaultSetter(ctx, variableValue, value)
+}
 
 // variable.Variable
 type BasicVariable struct {
-	getter GetterFunc
-	setter SetterFunc
+	getter *getterImpl
```
[the remark this answers is not in the document]
Comment (antJack) [reply pkg/variable/var.go:71]: BasicVariable.Getter()/Setter() methods need to return these fields as interface

**Review by wangfakang — APPROVED** [review]
LGTM.

**Review by nejisama — APPROVED** [review]
lgtm

**Pull request merged**